File: backtester/src/Strategy.cpp
#include "Strategy.h"

Strategy::~Strategy() = default;

void Strategy::on_start() {}

void Strategy::on_update(const MarketEvent&) {}

void Strategy::set_context(Context* context_) {
    context = context_;
}

void Strategy::send_signal(std::unique_ptr<Signal> signal) {
    if (context == nullptr || context->sink == nullptr) {
        return;
    }

    auto se = std::make_unique<SignalEvent>();
    se->signal = std::move(signal);
    context->sink->publish(std::move(se));
}

File: src/Strategy.py
from src.Types import *
from src.Events import MarketEvent
from src.Context import *

class Strategy(object):
    def __init__(self, name):
        self.name = name
        self.context: Context = None

    def on_start(self):
        pass

    def on_update(self, event: MarketEvent):
        pass

    #!TODO: Implement
    def send_signal(self, signal: Signal):
        pass

    def set_context(self, context: Context):
        self.context = context